src/runtime.rs

```diff
@@ -0,0 +1,161 @@
+// This file is Copyright its original authors, visible in version control history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. You may not use this file except in
+// accordance with one or both of these licenses.
+
+#[cfg(tokio_unstable)]
+use crate::logger::log_trace;
+use crate::logger::{log_error, LdkLogger, Logger};
+
+use tokio::task::JoinHandle;
+
+use std::fmt;
+use std::future::Future;
+use std::sync::{Arc, RwLock};
+
+pub(crate) struct Runtime {
+	state: RwLock<RuntimeState>,
+	logger: Arc<Logger>,
+}
+
+impl Runtime {
+	pub fn new(logger: Arc<Logger>) -> Self {
+		let state = RwLock::new(RuntimeState::Stopped);
+		Self { state, logger }
+	}
+
+	pub fn start(&self) -> Result<(), RuntimeError> {
+		let mut state_lock = self.state.write().unwrap();
+		if !matches!(*state_lock, RuntimeState::Stopped) {
+			return Err(RuntimeError::AlreadyRunning);
+		}
+		match tokio::runtime::Handle::try_current() {
+			Ok(handle) => *state_lock = RuntimeState::Handle(handle),
```

Comment:
Is it necessary to save this handle, or will - after starting - `tokio::runtime::Handle::try_current()` always return it?

Reply:
How about always using the runtime context only when needed and if it exists, and otherwise falling back to the handle that was provided to start or the newly instantiated runtime?

Reply:
What you suggest there is - I think - more predictable for the user? No hidden capture of a context.